lxd-agent/api_1.0.go

```diff
@@ -61,3 +67,89 @@ func api10Get(d *Daemon, r *http.Request) response.Response {
 
 	return response.SyncResponseETag(true, fullSrv, fullSrv)
 }
+
+func api10Put(d *Daemon, r *http.Request) response.Response {
+	var data api.API10Put
+
+	err := json.NewDecoder(r.Body).Decode(&data)
+	if err != nil {
+		return response.ErrorResponse(http.StatusInternalServerError, err.Error())
+	}
+
+	d.serverCID = data.CID
+	d.serverPort = data.Port
+	d.serverCertificate = data.Certificate
+
+	// Try connecting to LXD server.
+	client, err := getClient(int(d.serverCID), int(d.serverPort), d.serverCertificate)
+	if err != nil {
+		return response.ErrorResponse(http.StatusInternalServerError, err.Error())
+	}
+
+	server, err := lxd.ConnectLXDHTTP(nil, client)
+	if err != nil {
+		return response.ErrorResponse(http.StatusInternalServerError, err.Error())
+	}
+
+	defer server.Disconnect()
+
+	// Let LXD know, we were able to connect successfully.
+	d.chConnected <- struct{}{}
+
+	if data.Devlxd {
+		err = startDevlxdServer(d)
+	} else {
+		err = stopDevlxdServer()
+	}
+
+	if err != nil {
+		return response.ErrorResponse(http.StatusInternalServerError, err.Error())
+	}
+
+	return response.EmptySyncResponse
+}
+
+func startDevlxdServer(d *Daemon) error {
+	servers["devlxd"] = devLxdServer(d)
+
+	// Prepare the devlxd server.
+	devlxdListener, err := createDevLxdlListener("/dev")
+	if err != nil {
+		return err
+	}
+
+	// Start the devlxd listener.
+	go func() {
+		err := servers["devlxd"].Serve(devlxdListener)
+		if err != nil {
+			if !errors.Is(err, http.ErrServerClosed) {
```

Comment:
Why isn't this considered an error, could do with a comment I think, as its not obvious why this isn't a problem.

Reply:
When `Close()` is called, `Serve()` will return with an error. This isn't a problem , and shouldn't result in the `lxd-agent` stopping entirely.

Reply:
Hrm, I'm still not quite following this.

If `stopDevlxdServer()` has previously been called, and thus `servers["devlxd"].Close()` has been called, then later if we want to start up devlxd again we would call `startDevlxdServer()`. Are you saying that at this point, because `Close()` has previously been called that `Server()` will return an `http.ErrServerClosed` error?

If so then this seems like a problem as it won't allow us to start/stop/start the devlxd listener.

Reply:
Yes right, but I'm confused by your previous answer. Why when calling `Serve()` is it acceptable to ignore the `http.ErrServerClosed` error?

You previously said:

> When Close() is called, Serve() will return with an error. This isn't a problem , and shouldn't result in the lxd-agent stopping entirely.

But that confused me, as it implied that once `Close()` was called once all subsequent calls to `Serve()` will return `http.ErrServerClosed`, which suggested that once `stopDevlxdServer()` was called once couldn't call `startDevlxdServer()` and have the devlxd server start again.

I'm not so much interested in why you're ignoring it (as you've explained "If we returned an error in this case, lxd-agent would quit, although nothing bad happened") but more as to why "Serve just got notified that the server is now closed which is perfectly fine".

Why is it perfectly fine?

Reply:
> But that confused me, as it implied that once Close() was called once all subsequent calls to Serve() will return http.ErrServerClosed, which suggested that once stopDevlxdServer() was called once couldn't call startDevlxdServer() and have the devlxd server start again.

That's incorrect. After calling `stopDevlxdServer()` you can call `startDevlxdServer()`, and it will work. That's because we call `servers["devlxd"] = devLxdServer(d)` which allows us to start it again.

> Why is it perfectly fine?

Because calling `Close()` is perfectly fine. IMHO it's not so much an error as a notification that the server has been closed.

Reply:
It seems fishy to me. We shouldn't be getting an error like that `Serve()`, and I'd prefer that we don't paper over the issue.
The contract with `Serve()` should be that if it returns an error then the service wasn't started. If its truly not an issue then the `Serve()` function itself should be detecting that error and then returning nil rather than expecting the caller to understand the nuances.

Reply:
Unfortunately, that's not how `Serve` was implemented: ["Serve always returns a non-nil error."](https://pkg.go.dev/net/http@go1.18.3#Serve)

Reply:
OK right so not we're getting somewhere. I think what you're saying is that `ErrServerClosed` is returned after the `Server` function has been stopped by `Close()` after it had been previously running successfully?

Its a pretty unusual UX to always return an error, but as you say, not something we can change, so best just add a comment there explaining we are ignoring that particular error as it indicates the service was shutdown intentionally.

Reply:
> I think what you're saying is that ErrServerClosed is returned after the Server function has been stopped by Close() after it had been previously running successfully?

That's what I've been trying to say all along ;)

I'll add a comment.

Reply:
Yes I see that now, but the key thing you missed was that it was be returned after starting successfully and subsequently being closed. The way I was interpreting you was that a prior call to `Close()` was leaving the listener in a state that it couldn't be started again.